### ISSUE TITLE
Allow object as input of API functions

### ISSUE THREAD

**Issue by tpluscode** [types/npm-jsonld#7]
I see that the jsonld functions now take `JsonLd` as input. It is defined in a way which causes compilation to complain when any arbitrary JS object is passed.

jsonld.js works just fine this way and I think the `JsonLd` typing could be updated accordingly

**Comment by blakeembrey** [comment]
The tests are working with an "arbitrary object". Please elaborate on the issue. You can always use the type when defining the object.

**Comment by tpluscode** [comment]
The code work. The thing I'm seeing is `tsc` throwing errors like

> Error:(28, 28) TS2345: Argument of type '{ "@context": ... }' is not assignable to parameter of type 'JsonLdObject | JsonLdObject[] | string'.
>  Type '{ "@context": ... }' is not assignable to type 'string'.

**Comment by tpluscode** [comment]
Hm I tried to get rid of that error. I think that adding `Object` explicitly works

```
export type JsonLd = JsonLdObject | JsonLdObject[] | Object | Object[];
```

What do you think?

**Comment by felixfbecker** [comment]
@tpluscode Please show the code where your error occurs. `JsonLdObject` is an interface, and whatever object is compatible with that interface can get passed to the function.

**Comment by tpluscode** [comment]
I'm seeing two cases. One where the input is an object created with an object literal: https://github.com/wikibus/heracles/blob/master/tests/ApiDocumentation-spec.ts#L17

And second where I'm calling jsonld on self (https://github.com/wikibus/heracles/blob/master/src/Resources.ts#L53) in which case the error reads 

> Error:(53, 31) TS2345: Argument of type 'this' is not assignable to parameter of type 'JsonLdObject | JsonLdObject[] | string'.
>  Type 'Resource' is not assignable to type 'JsonLdObject | JsonLdObject[] | string'.
>  Type 'Resource' is not assignable to type 'string'.
>  Type 'this' is not assignable to type 'string'.
>  Type 'Resource' is not assignable to type 'string'.

**Comment by felixfbecker** [comment]
It's hard to tell on GitHub because I don't want to clone the repo and you don't use explicit type annotations, but one of your properties in `Resource` is not matching the type of the interface, so it appears to me that the typings are correct here and the usage is wrong. Putting `Object` in there would cripple the type safety and allow anything (except primitives).

**Comment by tpluscode** [comment]
I added return types to `Resource` and it didn't help. Which property do you think does not match the interface? There are three return types: `string`, `boolean` and `string | string[]`.

**Comment by tpluscode** [comment]
And what about imported var? I see that usage like below works fine.

```
jsonld.compact({
  '@id:' '...',
  '@context': { ... }
});
```

But if that object is imported variable, tsc complains

**Comment by felixfbecker** [comment]
Are you saying this fails while your example works?

``` ts
// a.ts
export const whatever = { '@id': '...', '@context': { ... } }
// b.ts
import {whatever} from './a;
jsonId.compact(whatever)
```

that is impossible, it is only the type that matters. The input needs to match the interface ;)

**Comment by tpluscode** [comment]
In my case `whatever` is actually inside a namespace. But yes, either way it doesn't work. I mean, it works. It just refuses to compile without an error :)

**Comment by felixfbecker** [comment]
Could you try to create an [SSCCE](http://www.sscce.org/)?

**Comment by tpluscode** [comment]
Will do in the evening.

**Comment by blakeembrey** [comment]
I think I see the issue, and it is probably because `jsonld` uses "own properties" and your code has methods as part of the interface, but functions are incompatible with the signature. To be honest, I don't think TypeScript can express this specific limitation. If `jsonld` (the library) can accept functions as "own properties", I'd be happy to add those to the input object so this would work.

**Comment by felixfbecker** [comment]
Methods are non-enumerable by default though

**Comment by blakeembrey** [comment]
In TypeScript or JavaScript? In any case, they're on the prototype but the `jsonld` library uses `Object.keys` internally, so never encounters the functions. That's all. Also, methods are enumerable in the JavaScript emit of TypeScript (http://www.typescriptlang.org/play/#src=class%20Foo%20%7B%0A%20%20%20%20method()%20%7B%20%7D%0A%7D%0A%0Afor%20(var%20key%20in%20new%20Foo())%20%7B%0A%20%20%20%20console.log(key)%0A%7D). TypeScript has no concept of prototypes here, so we can't type an interface that checks "own properties" only.

**Comment by tpluscode** [comment]
Guys, would you still like to see the SSCCE?

**Comment by blakeembrey** [comment]
However, it seems the above comment is irrelevant. I tested it with the below snippet and it worked, so we can probably update the interface accordingly:

``` js
require('jsonld').compact({ '@context': { schema: 'http://schema.org/' }, 'schema:foo': function () {} }, {}, console.log)
```

Edit: It just stringifies the function input, which I think is probably unwanted. We can keep the interface as is or change it then, but I like the idea of warning on bad input which is what TypeScript is designed for.

**Comment by felixfbecker** [comment]
I agree that this is probably unwanted. You should pick the attributes out of the object.

**Comment by tpluscode** [comment]
I've finally created that example. You can compile files from this gist: https://gist.github.com/tpluscode/d962f5bb894392dd6f7887f07a2d5ca7
